src/kOS.Safe/Encapsulation/ListValue.cs

```diff
@@ -141,141 +141,78 @@ public override bool SetSuffix(string suffixName, object value)
-        /// will break:
-        /// <returns>long string including eolns, ready for printing</returns>
-        private string ListDumpDeep()
-        {
-            maxVerboseDepth = 99;
-            return ListDump();
+            if (!(index is int)) throw new KOSException("The index must be an integer number");
+
+            internalList[(int)index] = (T)value;
         }
 
-        /// <summary>
-        /// This is the engine underneath ListDump shallow/deep.
-        /// </summary>
-        /// <returns>string dump of the list</returns>
-        private string ListDump()
+        public string[] Dump(int limit, int depth = 0)
         {
-            const int SPACES_PER_INDENT = 2;
+            var toReturn = new List<string>();
 
-            ++currentNestDepth;
-            bool truncateHere = currentNestDepth > maxVerboseDepth;
+            var listString = string.Format("LIST of {0} items", Count);
+            toReturn.Add(listString);
 
-            if (truncateHere)
-            {
-                --currentNestDepth;
-                return TerseDump();
-            }
-            var contents = new StringBuilder();
-            contents.AppendLine(TerseDump() + ":");
-            var indent = new string(' ', currentNestDepth * SPACES_PER_INDENT);
-            for (int i = 0; i < internalList.Count; ++i)
+            if (limit <= 0) return toReturn.ToArray();
+
+            for (int index = 0; index < internalList.Count; index++)
             {
-                contents.AppendLine(string.Format("{0}[{1,2}]= {2}", indent, i, internalList[i]));
-            }
-            --currentNestDepth;
-            return contents.ToString();
-        }
+                var item = internalList[index];
 
-        public static ListValue<T> CreateList<TU>(IEnumerable<TU> list)
-        {
-            return new ListValue<T>(list.Cast<T>());
-        }
+                var dumper = item as IDumper;
+                if (dumper != null)
+                {
+                    var entry = string.Empty.PadLeft(depth * INDENT_SPACES);
 
-        public object GetIndex(int index)
-        {
-            return internalList[index];
+                    var itemDump = dumper.Dump(limit - 1, depth + 1);
+
+                    var itemString = string.Format("  [{0,2}]= {1}", index, itemDump[0]);
+                    entry += itemString;
+
+                    toReturn.Add(entry);
+
+                    for (int i = 1; i < itemDump.Length; i++)
+                    {
+                        var subEntry = string.Format("{0}", itemDump[i]);
+                        toReturn.Add(subEntry);
+                    }
+                }
+                else
+                {
+                    var entry = string.Empty.PadLeft(depth * INDENT_SPACES);
+                    entry += string.Format("  [{0,2}]= {1}", index, item);
+                    toReturn.Add(entry); 
+                }
+            }
+            return toReturn.ToArray();
```

Comment:
Wow, Github's diff'er really made a mess of that stuff above this point.  I can't follow at all what happened.  I'm going to have to wait until you're done and I pull this and look at it locally with smarter diffing tools to decipher what changed here.